source/app/controllers/urls_controller.rb

```diff
@@ -1,2 +1,26 @@
 class UrlsController < ApplicationController
+  include UrlsHelper
+
+  def index
+  	@urls = Url.all
+  	@url = Url.new(address: "https://")
+  end
+
+  def create
+  	@url = Url.new(url_params)
+  	@url.click_count = 0
+  	if @url.save
+  	  redirect_to root_path
+  	else
+  	  @urls = Url.all
+  	  render 'index'
+  	end
+  end
+
+  def show
+  	@url = Url.find_by(slug: params[:slug])
+  	@url.click_count += 1
+  	@url.save
```

Comment:
ActiveRecord [has a method for this sort of thing](http://api.rubyonrails.org/classes/ActiveRecord/Persistence.html#method-i-increment).

There's a subtle bug as currently written. Do you see it?

Reply:
Was it that there was no error handling for invalid slugs?

Reply:
No, concurrent requests for the same slug would undercount the clicks.

Reply:
Would you recommend locking? Or is there a better way to do this?

Reply:
Oops. I thought the increment method did an atomic increment in the database. Looking at its implementation now, I see that it does the same thing your code did here: load the value into memory, add one, then save the new value. 

A lock would certainly resolve this (and ActiveRecord has a with_lock method to make it easy), but I prefer a SQL update command without a lock.

Reply:
I discovered the class method :increment_counter which does an atomic increment of the field given the id and have implemented it in my latest commit. Thanks for showing me this as I now have some understanding of concurrency issues and atomicity

Reply:
Perfect!